From d4b7a8fd28ec7704f750199b6b695d3500c36843 Mon Sep 17 00:00:00 2001
From: Yangbo Lu <yangbo.lu@nxp.com>
Date: Fri, 28 Dec 2018 11:07:57 +0800
Subject: [PATCH 450/706] ptp_qoriq: move tmr_tevent reset after timestamp FIFO
 cleaning

The tmr_tevent should be reset after timestamp FIFO cleaning.
Otherwise external trigger event bits couldn't be reset.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>
(cherry picked from commit 39253708f1820ef7759cdf3df586fb5eef7a483f)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/ptp/ptp_qoriq.c | 24 +++++++++++-------------
 1 file changed, 11 insertions(+), 13 deletions(-)

diff --git a/drivers/ptp/ptp_qoriq.c b/drivers/ptp/ptp_qoriq.c
index 5ebd6e64bf5d..e0730eac44b3 100644
--- a/drivers/ptp/ptp_qoriq.c
+++ b/drivers/ptp/ptp_qoriq.c
@@ -308,11 +308,6 @@ int ptp_qoriq_enable(struct ptp_clock_info *ptp,
 	unsigned long flags;
 	u32 bit, event, mask, lo, hi;
 
-	spin_lock_irqsave(&qoriq_ptp->lock, flags);
-	event = qoriq_read(qoriq_ptp, &regs->ctrl_regs->tmr_tevent);
-	qoriq_write(qoriq_ptp, &regs->ctrl_regs->tmr_tevent, event);
-	spin_unlock_irqrestore(&qoriq_ptp->lock, flags);
-
 	switch (rq->type) {
 	case PTP_CLK_REQ_EXTTS:
 		switch (rq->extts.index) {
@@ -325,6 +320,7 @@ int ptp_qoriq_enable(struct ptp_clock_info *ptp,
 		default:
 			return -EINVAL;
 		}
+
 		spin_lock_irqsave(&qoriq_ptp->lock, flags);
 		extts_read_clean(qoriq_ptp, rq->extts.index, &lo, &hi);
 		mask = qoriq_read(qoriq_ptp, &regs->ctrl_regs->tmr_temask);
@@ -332,10 +328,8 @@ int ptp_qoriq_enable(struct ptp_clock_info *ptp,
 			mask |= bit;
 		else
 			mask &= ~bit;
-		qoriq_write(qoriq_ptp, &regs->ctrl_regs->tmr_temask, mask);
 		spin_unlock_irqrestore(&qoriq_ptp->lock, flags);
-		return 0;
-
+		break;
 	case PTP_CLK_REQ_PPS:
 		spin_lock_irqsave(&qoriq_ptp->lock, flags);
 		mask = qoriq_read(qoriq_ptp, &regs->ctrl_regs->tmr_temask);
@@ -343,15 +337,19 @@ int ptp_qoriq_enable(struct ptp_clock_info *ptp,
 			mask |= PP1EN;
 		else
 			mask &= ~PP1EN;
-		qoriq_write(qoriq_ptp, &regs->ctrl_regs->tmr_temask, mask);
 		spin_unlock_irqrestore(&qoriq_ptp->lock, flags);
-		return 0;
-
-	default:
 		break;
+	default:
+		return -EOPNOTSUPP;
 	}
 
-	return -EOPNOTSUPP;
+	spin_lock_irqsave(&qoriq_ptp->lock, flags);
+	event = qoriq_read(qoriq_ptp, &regs->ctrl_regs->tmr_tevent);
+	qoriq_write(qoriq_ptp, &regs->ctrl_regs->tmr_tevent, event);
+	qoriq_write(qoriq_ptp, &regs->ctrl_regs->tmr_temask, mask);
+	spin_unlock_irqrestore(&qoriq_ptp->lock, flags);
+
+	return 0;
 }
 
 static const struct ptp_clock_info ptp_qoriq_caps = {
-- 
2.17.1

